synapse/storage/databases/main/room.py

```diff
@@ -1311,6 +1273,64 @@ async def get_join_event_id_and_device_lists_stream_id_for_partial_state(
         )
         return result["join_event_id"], result["device_lists_stream_id"]
 
+    def get_un_partial_stated_rooms_token(self) -> int:
+        # TODO(multiple writers): This is inappropriate if there are multiple
+        #     writers because workers that don't write often will hold all
+        #     readers up.
```

Comment:
I thought it's slow workers holding uncompleted stream IDs that hold readers up, rather than infrequent writers?

Reply:
First thought: I've added a reference so this is explained in the future, but on the multi-writer ID generator, `get_persisted_upto_position` says:

> Get the max position where all previous positions have been
        persisted.
> 
> Note: In the worst case scenario this will be equal to the minimum
        position across writers. This means that the returned position here can
        lag if one writer doesn't write very often.

As far as I understand, the correct way to fix this would be for readers to track the seen position of each writer, like we do (sometimes) in sync tokens.


Second thought: you sound right ... maybe `get_persisted_upto_position` is actually wrong.

Third thought: Thinking a bit, I believe the problem might arise when a writers generates a stream ID but doesn't use it (e.g. if the transaction fails). In this case, there'll be a gap left in the stream and other workers won't understand that the writer has abandoned writing it until that same writer writes a new row. The current 'persisted up to position' appears to be more or less 'the position of the first gap after the current `min(writer_positions)`, minus 1', so if you have
`{A: 5, B: 6, C:7, D: 9}` then the persisted position would be `7` (the min is `5` but as we have `6` and `7`, the first gap is `8`).

(`MultiWriterIdGenerator` could really use a longer-form piece of documentation explaining the principles of operation, I think...)

Reply:
Curious, if that's what the docstring for `get_persisted_upto_position` says, then I'm inclined to (pretend to) believe it.

I would have thought that we handle both successful and failed transactions in the same way due to the exit logic in the context manager that the multi-writer ID generator returns. Though it seems that replication isn't handled automatically and someone needs to call `advance(instance_name, stream_id)` after receiving a replication row.

It's all somewhat mysterious and I'm happy to leave the comment alone.

Reply:
Possibly the hole you don't notice is that you need to handle a writer crashing half way through as well ... so maybe it's much more theoretical than it sounds? Hrm. One to revisit at the right time, I think.